### PR TITLE
Implement new plan optimizer: PredicatePushDown

### PR DESCRIPTION
**Pull request by liuminghui233**
In this PR, I implement new plan optimizer: PredicatePushDown.

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/expression/visitor/predicate/ConvertPredicateToTimeFilterVisitor.java [hunk 1]
```diff
@@ -88,17 +88,17 @@ private Expression rewriteInExpressionToEqual(InExpression inExpression) {
 
   @Override
   public Filter visitIsNullExpression(IsNullExpression isNullExpression, Void context) {
-    throw new UnsupportedOperationException("TIMESTAMP does not support IS NULL");
+    throw new UnsupportedOperationException("TIMESTAMP does not support IS NULL/IS NOT NULL");
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/expression/visitor/predicate/ConvertPredicateToTimeFilterVisitor.java:91]: combine with ConvertPredicateToFilterVisitor

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/optimization/PredicatePushDown.java [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+      }
+      return new TransformNode(
+          context.genPlanNodeId(),
+          resultNode,
+          outputExpressions,
+          pushDownFilterNode.isKeepNull(),
+          pushDownFilterNode.getScanOrder());
+    }
+
+    private PlanNode planProject(PlanNode resultNode, RewriterContext context) {
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/optimization/PredicatePushDown.java:342]: comment

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/optimization/PredicatePushDown.java [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+        // only support push down to aligned scan
+        return node;
+      }
+
+      Expression inheritedPredicate = context.getInheritedPredicate();
+      List<Expression> conjuncts = PredicateUtils.extractConjuncts(inheritedPredicate);
+
+      List<PlanNode> children = node.getChildren();
+
+      List<List<Expression>> pushDownConjunctsForEachChild = new ArrayList<>(children.size());
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/optimization/PredicatePushDown.java]: done

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/optimization/PredicatePushDown.java [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+      boolean enablePushDown = context.isEnablePushDown();
+      context.reset();
+
+      if (enablePushDown) {
+        return rewrittenChild;
+      }
+      return node;
+    }
+
+    private boolean fromHaving(Expression predicate) {
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/optimization/PredicatePushDown.java]: done

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/PredicateUtils.java [hunk 5]
```diff
@@ -365,4 +330,17 @@ private static void extractConjuncts(Expression predicate, Set<Expression> conju
       conjuncts.add(predicate);
     }
   }
+
+  public static boolean predicateCanPushDownToSource(
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/PredicateUtils.java]: done

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/PredicateUtils.java [hunk 6]
```diff
@@ -365,4 +330,17 @@ private static void extractConjuncts(Expression predicate, Set<Expression> conju
...
+      Expression predicate, PartialPath checkedSourcePath, boolean isBuildPlanUseTemplate) {
+    return new PredicateCanPushDownToSourceChecker()
+        .process(
+            predicate,
+            new PredicateCanPushDownToSourceChecker.Context(
+                checkedSourcePath, isBuildPlanUseTemplate));
+  }
+
+  public static boolean predicateCanPushIntoScan(Expression predicate) {
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/PredicateUtils.java:342]: done

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/optimization/PredicatePushDown.java (hunk 2 above)
Comment (liuminghui233) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/optimization/PredicatePushDown.java:330]: done

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/common/MPPQueryContext.java [hunk 7]
```diff
@@ -62,6 +65,8 @@ public class MPPQueryContext {
 
   private int acquiredLockNum;
 
+  private final Set<NodeRef<FilterNode>> fromWhereFilterNodes = new HashSet<>();
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/common/MPPQueryContext.java]: put this into Analysis

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**